File: ironworks/art/characters.py
"""Character sprite sheets.

Every person - the player, hired staff, townsfolk - comes off the same rig with
different palettes, uniforms, headgear and hand props, so a crowd reads as
individuals without drifting out of the game's visual language.
"""
import math
import pygame

from ..core.rng import Rng
from ..core.utils import TAU
from . import draw as D
from . import palette as P
from .palette import mix, shade, rgba

FW, FH = 34, 46      # frame size
COLS, ROWS = 8, 4    # 0 idle | 1..5 walk | 6..7 work   x   dir N/E/S/W

SKIN = [(240, 200, 160), (224, 176, 136), (201, 145, 103),
        (169, 115, 76), (132, 83, 52), (95, 58, 34)]
HAIR = [(43, 29, 19), (74, 47, 27), (122, 75, 34), (176, 133, 68),
        (216, 207, 192), (26, 26, 30), (107, 58, 42)]

ROLE_STYLE = {
    'player':    dict(shirt=(200, 98, 42), pants=(58, 68, 80), helmet=P.YELLOW, vest=True, tool='wrench'),
    'worker':    dict(shirt=(77, 106, 134), pants=(57, 66, 77), helmet=P.YELLOW, vest=True, tool='box'),
    'engineer':  dict(shirt=(216, 210, 196), pants=(57, 66, 77), helmet=(232, 226, 212), vest=True, tool='tablet', glasses=True),
    'mechanic':  dict(shirt=(90, 95, 70), pants=(59, 63, 48), helmet=(138, 90, 42), vest=False, tool='wrench'),
    'logistics': dict(shirt=(63, 122, 82), pants=(57, 66, 77), helmet=P.ORANGE, vest=True, tool='box'),
    'manager':   dict(shirt=(46, 58, 74), pants=(36, 44, 54), helmet=None, vest=False, tool='tablet', tie=True),
    'researcher': dict(shirt=(230, 233, 236), pants=(74, 85, 96), helmet=None, vest=False, tool='tablet', coat=True, glasses=True),
    'technician': dict(shirt=(106, 79, 134), pants=(57, 66, 77), helmet=(191, 198, 205), vest=True, tool='wrench'),
    'civilian':  dict(shirt=None, pants=None, helmet=None, vest=False, tool=None),
    'vendor':    dict(shirt=(138, 90, 42), pants=(74, 58, 42), helmet=None, vest=False, tool='box', apron=True),
}

_sheets = {}
_portraits = {}


def character_sheet(role='worker', seed=1):
    key = (role, seed % 12)
    hit = _sheets.get(key)
    if hit is not None:
        return hit
    rnd = Rng(4242 + seed * 7919 + len(role) * 37)
    base = ROLE_STYLE.get(role, ROLE_STYLE['worker'])
    look = dict(
        skin=SKIN[int(rnd() * len(SKIN)) % len(SKIN)],
        hair=HAIR[int(rnd() * len(HAIR)) % len(HAIR)],
        shirt=base['shirt'] or [(138, 90, 74), (74, 106, 138), (106, 122, 74),
                                (122, 74, 106), (138, 122, 74)][int(rnd() * 5) % 5],
        pants=base['pants'] or [(58, 63, 74), (74, 64, 56), (47, 58, 68)][int(rnd() * 3) % 3],
        helmet=base['helmet'], vest=base['vest'], tool=base['tool'],
        tie=base.get('tie', False), coat=base.get('coat', False),
        glasses=base.get('glasses', False) or rnd() < 0.15,
        apron=base.get('apron', False),
        beard=rnd() < 0.3, long_hair=rnd() < 0.35, boots=(42, 38, 34),
    )

    def paint(s, w, h, k, look=look):
        for d in range(ROWS):
            for c in range(COLS):
                _frame(s, look, d, c, k, c * FW * k, d * FH * k)

    sheet = D.render_sprite(FW * COLS, FH * ROWS, paint)
    if pygame.display.get_surface():
        sheet = sheet.convert_alpha()
    out = dict(surf=sheet, fw=FW, fh=FH, look=look)
    _sheets[key] = out
    return out


def _frame(s, L, direction, col, k, ox, oy):
    cx = ox + FW * k / 2
    feet = oy + (FH - 5) * k
    walk = 1 <= col <= 5
    work = col >= 6
    phase = (col - 1) / 5 if walk else 0
    swing = math.sin(phase * TAU) if walk else 0
    bob = abs(math.cos(phase * TAU)) * 1.4 * k if walk else 0
    work_pose = -1 if not work else (0 if col == 6 else 1)
    top = feet - 32 * k - bob
    side = 1 if direction == 1 else (-1 if direction == 3 else 0)

    D.ellipse(s, rgba((8, 10, 14), 0.30), cx, feet + k, 8.5 * k, 3.4 * k)

    # ---- legs
    leg_swing = swing * 4.2 * k
    for sgn in (-1, 1):
        off = sgn * (1.2 * k if side else 3.1 * k)
        fwd = leg_swing * sgn * side if side else leg_swing * sgn
        D.rr(s, (cx + off - 2.6 * k + (fwd * 0.6 if side else 0), top + 17 * k, 5.2 * k, 12 * k),
             L['pants'] if sgn == 1 else shade(L['pants'], -0.18), int(2.2 * k))
        bx = cx + off - 3 * k + (fwd * 0.8 if side else 0)
        D.rr(s, (bx, feet - 4 * k, 6 * k, 4.4 * k), L['boots'], int(1.8 * k))
        D.rect(s, rgba(P.WHITE, 0.12), (bx, feet - 4 * k, 6 * k, k))

    # ---- torso
    tw = 9.5 * k if side else 12 * k
    th = 17 * k
    tx, ty = cx - tw / 2, top + 2 * k
    D.grad_rect(s, (tx, ty, tw, th), [(0.0, shade(L['shirt'], 0.2)), (0.55, L['shirt']),
                                      (1.0, shade(L['shirt'], -0.28))], radius=int(3 * k))
    D.rr(s, (tx, ty, tw, th), rgba(P.BLACK, 0.35), int(3 * k), 1)

    if L['coat']:
        D.rr(s, (tx - k, ty + k, tw + 2 * k, th + 4 * k), rgba((242, 244, 246), 0.95), int(3 * k))
        if direction == 2:
            D.line(s, rgba((174, 181, 188), 0.9), (cx, ty + 2 * k), (cx, ty + th + 3 * k), 1)
    if L['apron']:
        D.rr(s, (tx + k, ty + 5 * k, tw - 2 * k, th - 3 * k), rgba((107, 74, 42), 0.9), int(2 * k))
    if L['vest']:
        D.rr(s, (tx - 0.8 * k, ty + 3 * k, tw + 1.6 * k, th - 3 * k), rgba((232, 226, 74), 0.92), int(2.4 * k))
        D.rect(s, rgba((216, 221, 226), 0.95), (tx - 0.8 * k, ty + 7 * k, tw + 1.6 * k, 2 * k))
        D.rect(s, rgba((216, 221, 226), 0.95), (tx - 0.8 * k, ty + 12 * k, tw + 1.6 * k, 2 * k))
        if direction == 2:
            D.rect(s, rgba((138, 132, 32), 0.5), (cx - 0.5 * k, ty + 3 * k, k, th - 3 * k))
    if L['tie'] and direction == 2:
        D.polygon(s, (166, 58, 46), [(cx, ty + k), (cx + 1.8 * k, ty + 4 * k),
                                     (cx, ty + 12 * k), (cx - 1.8 * k, ty + 4 * k)])

    # ---- arms (angles in radians, 0 = straight down)
    for sgn in (-1, 1):
        ax = cx + sgn * (tw / 2 + 1.2 * k)
        ay = ty + 3 * k
        if work_pose >= 0:
            ang = sgn * (-0.9 if work_pose else -0.5)
            ay -= k
        elif side:
            ang = -swing * 0.5 * (1 if sgn == side else -1)
        else:
            ang = sgn * (0.1 + swing * 0.12)
        ln = 11 * k
        ex = ax + math.sin(ang) * ln * 0.6
        ey = ay + math.cos(ang) * ln * 0.6
        D.line(s, (242, 244, 246) if L['coat'] else shade(L['shirt'], -0.1 if sgn == 1 else 0.08),
               (ax, ay), (ex, ey), max(2, int(4.2 * k)))
        hx = ax + math.sin(ang) * ln
        hy = ay + math.cos(ang) * ln
        D.line(s, L['skin'], (ex, ey), (hx, hy), max(2, int(3.8 * k)))

    # ---- head
    hy = top - 2.5 * k
    hr = 5.6 * k
    D.ellipse(s, L['skin'], cx, hy, hr, hr * 1.06)
    D.ellipse(s, rgba(P.BLACK, 0.10), cx + (side * 1.6 * k if side else 0), hy + 1.6 * k, hr * 0.9, hr * 0.7)
    D.rect(s, rgba(P.BLACK, 0.22), (cx - 2.2 * k, hy + hr * 0.8, 4.4 * k, 2 * k))

    if not L['helmet'] or direction == 0:
        D.arc(s, L['hair'], cx, hy - 0.8 * k, hr * 1.02, 0, math.pi, max(2, int(hr * 0.9)))
        if L['long_hair']:
            D.rect(s, L['hair'], (cx - hr, hy - k, hr * 2, 5 * k))

    if direction == 2:
        D.circle(s, (42, 33, 24), cx - 2 * k, hy + 0.4 * k, max(1, 0.8 * k))
        D.circle(s, (42, 33, 24), cx + 2 * k, hy + 0.4 * k, max(1, 0.8 * k))
        D.arc(s, rgba((107, 70, 54), 0.8), cx, hy + 2.6 * k, 1.5 * k, 0.2, math.pi - 0.2, max(1, int(k)))
        if L['beard']:
            D.arc(s, rgba(L['hair'], 0.85), cx, hy + 2.4 * k, 3.4 * k, 0.15, math.pi - 0.15, max(1, int(2 * k)))
        if L['glasses']:
            D.circle(s, rgba(P.GLASS, 0.3), cx - 2 * k, hy + 0.4 * k, 1.8 * k)
            D.circle(s, rgba(P.GLASS, 0.3), cx + 2 * k, hy + 0.4 * k, 1.8 * k)
            D.circle(s, rgba((28, 28, 32), 0.9), cx - 2 * k, hy + 0.4 * k, 1.9 * k, 1)
            D.circle(s, rgba((28, 28, 32), 0.9), cx + 2 * k, hy + 0.4 * k, 1.9 * k, 1)
    elif side:
        D.circle(s, (42, 33, 24), cx + side * 2.4 * k, hy + 0.4 * k, max(1, 0.75 * k))

    if L['helmet'] and direction != 0:
        D.arc(s, L['helmet'], cx, hy - 1.8 * k, hr * 1.08, 0, math.pi, max(2, int(hr * 0.85)))
        D.rect(s, L['helmet'], (cx - hr * 1.08, hy - 2.6 * k, hr * 2.16, 1.6 * k))
        bx = cx + (side * 1.5 * k if side else 0)
        D.rr(s, (bx - hr * 1.12, hy - 1.8 * k, hr * 2.24, 1.7 * k), shade(L['helmet'], -0.18), int(k))
        D.rr(s, (bx - hr * 1.12, hy - 1.8 * k, hr * 2.24, 1.7 * k), rgba(P.BLACK, 0.4), int(k), 1)
        if direction == 2:
            D.rect(s, rgba(P.WHITE, 0.22), (cx - 0.7 * k, hy - hr * 1.05 - 1.8 * k, 1.4 * k, hr * 0.9))
    elif L['helmet']:
        D.circle(s, shade(L['helmet'], -0.05), cx, hy, hr * 1.08)
        D.circle(s, rgba(P.WHITE, 0.2), cx - 1.6 * k, hy - 1.6 * k, hr * 0.4)

    # ---- carried tool
    if L['tool'] and (work_pose >= 0 or col == 0):
        if L['tool'] == 'wrench':
            hx = cx + (side * 7 * k if side else 7 * k)
            hy2 = top + 13 * k
            tilt = -0.5 if work_pose == 1 else 0.25
            ex = hx + math.sin(tilt) * 9 * k
            ey = hy2 - math.cos(tilt) * 9 * k
            D.line(s, P.STEEL_HI, (hx, hy2), (ex, ey), max(1, int(2 * k)))
            D.circle(s, P.STEEL_MID, ex, ey - k, 2.6 * k)
            D.circle(s, (26, 28, 32), ex, ey - 1.6 * k, 1.2 * k)
        elif L['tool'] == 'box':
            D.rr(s, (cx - 6 * k, top + 8 * k, 12 * k, 9 * k), (165, 128, 74), int(1.5 * k))
            D.rr(s, (cx - 6 * k, top + 8 * k, 12 * k, 9 * k), (74, 55, 23), int(1.5 * k), 1)
            D.line(s, (74, 55, 23), (cx - 6 * k, top + 12.5 * k), (cx + 6 * k, top + 12.5 * k), 1)
        elif L['tool'] == 'tablet':
            px = cx + (side * 5 * k if side else 5 * k)
            D.rr(s, (px - 3.5 * k, top + 10 * k, 7 * k, 5 * k), (42, 47, 54), int(k))
            D.rect(s, rgba(P.GLOW_COLD, 0.8), (px - 2.8 * k, top + 10.6 * k, 5.6 * k, 3.8 * k))


def portrait(role, seed, size=56):
    """Framed 3/4 bust for dialogue and the employee roster."""
    key = (role, seed % 12, size)
    hit = _portraits.get(key)
    if hit is not None:
        return hit
    sheet = character_sheet(role, seed)
    s = D.surf(size, size)
    D.grad_rect(s, (0, 0, size, size), [(0.0, (51, 58, 68)), (1.0, (28, 33, 40))], radius=6)
    scale = size / 26.0
    body = sheet['surf'].subsurface(pygame.Rect(2 * FW, 2 * FH, FW, FH))
    body = pygame.transform.smoothscale(body, (int(FW * scale), int(FH * scale)))
    clip = D.surf(size, size)
    # frame the head and shoulders: the head sits ~14px down a 46px frame
    clip.blit(body, (int(size / 2 - body.get_width() / 2), int(size * 0.46 - 14 * scale)))
    mask = D.surf(size, size)
    D.rr(mask, (0, 0, size, size), (255, 255, 255, 255), 6)
    clip.blit(mask, (0, 0), special_flags=pygame.BLEND_RGBA_MULT)
    s.blit(clip, (0, 0))
    D.rr(s, (1, 1, size - 2, size - 2), rgba((13, 15, 18), 0.9), 6, 2)
    D.rr(s, (2.5, 2.5, size - 5, size - 5), rgba(P.UI_TRIM, 0.6), 5, 1)
    if pygame.display.get_surface():
        s = s.convert_alpha()
    _portraits[key] = s
    return s

File: src/art/characters.js
// Character sprite sheets. Every person in the game (player, workers, townsfolk) is
// drawn from the same rig with different palettes, uniforms, headgear and props, so
// the crowd looks individual but consistent.

import { makeCanvas, rr, glow, dropShadow } from './draw.js';
import { P, rgba, shade, mix } from './palette.js';
import { mulberry32 } from '../core/rng.js';
import { TAU } from '../core/utils.js';

export const FW = 34, FH = 46;           // frame size
export const COLS = 8;                   // 0 idle | 1..5 walk | 6..7 work
export const ROWS = 4;                   // dir: 0 N, 1 E, 2 S, 3 W

const SKIN = ['#f0c8a0', '#e0b088', '#c99167', '#a9734c', '#845334', '#5f3a22'];
const HAIR = ['#2b1d13', '#4a2f1b', '#7a4b22', '#b08544', '#d8cfc0', '#1a1a1e', '#6b3a2a'];

export const ROLE_STYLE = {
  player:    { shirt: '#c8622a', pants: '#3a4450', helmet: P.paintYellow, vest: true, tool: 'wrench' },
  worker:    { shirt: '#4d6a86', pants: '#39424d', helmet: P.paintYellow, vest: true, tool: 'box' },
  engineer:  { shirt: '#d8d2c4', pants: '#39424d', helmet: '#e8e2d4', vest: true, tool: 'tablet', glasses: true },
  mechanic:  { shirt: '#5a5f46', pants: '#3b3f30', helmet: '#8a5a2a', vest: false, tool: 'wrench' },
  logistics: { shirt: '#3f7a52', pants: '#39424d', helmet: P.paintOrange, vest: true, tool: 'box' },
  manager:   { shirt: '#2e3a4a', pants: '#242c36', helmet: null, vest: false, tool: 'tablet', tie: true },
  researcher:{ shirt: '#e6e9ec', pants: '#4a5560', helmet: null, vest: false, tool: 'tablet', coat: true, glasses: true },
  technician:{ shirt: '#6a4f86', pants: '#39424d', helmet: '#bfc6cd', vest: true, tool: 'wrench' },
  civilian:  { shirt: null, pants: null, helmet: null, vest: false, tool: null },
  vendor:    { shirt: '#8a5a2a', pants: '#4a3a2a', helmet: null, vest: false, tool: 'box', apron: true },
};

const sheetCache = new Map();

export function characterSheet(role = 'worker', seed = 1) {
  const key = role + ':' + (seed % 12);
  if (sheetCache.has(key)) return sheetCache.get(key);
  const rnd = mulberry32(4242 + seed * 7919 + role.length * 37);
  const base = ROLE_STYLE[role] || ROLE_STYLE.worker;
  const look = {
    skin: SKIN[(rnd() * SKIN.length) | 0],
    hair: HAIR[(rnd() * HAIR.length) | 0],
    shirt: base.shirt || ['#8a5a4a', '#4a6a8a', '#6a7a4a', '#7a4a6a', '#8a7a4a'][(rnd() * 5) | 0],
    pants: base.pants || ['#3a3f4a', '#4a4038', '#2f3a44'][(rnd() * 3) | 0],
    helmet: base.helmet,
    vest: base.vest, tool: base.tool, tie: base.tie, coat: base.coat,
    glasses: base.glasses || rnd() < 0.15,
    apron: base.apron,
    beard: rnd() < 0.3, long: rnd() < 0.35,
    boots: '#2a2622',
  };
  const { c, g } = makeCanvas(FW * COLS, FH * ROWS);
  for (let dir = 0; dir < ROWS; dir++)
    for (let col = 0; col < COLS; col++) {
      g.save();
      g.translate(col * FW, dir * FH);
      drawFrame(g, look, dir, col);
      g.restore();
    }
  const sheet = { canvas: c, fw: FW, fh: FH, look };
  sheetCache.set(key, sheet);
  return sheet;
}

function drawFrame(g, L, dir, col) {
  const cx = FW / 2, feet = FH - 5;
  const walk = col >= 1 && col <= 5;
  const work = col >= 6;
  const phase = walk ? (col - 1) / 5 : 0;
  const swing = walk ? Math.sin(phase * TAU) : 0;
  const bob = walk ? Math.abs(Math.cos(phase * TAU)) * 1.4 : 0;
  const workPose = work ? (col === 6 ? 0 : 1) : -1;

  // ---- ground shadow
  g.save();
  g.fillStyle = 'rgba(8,10,14,0.30)';
  g.beginPath(); g.ellipse(cx, feet + 1, 8.5, 3.4, 0, 0, TAU); g.fill();
  g.restore();

  const top = feet - 32 - bob;     // shoulder line
  const side = dir === 1 ? 1 : dir === 3 ? -1 : 0;

  // ---- legs
  const legSwing = swing * 4.2;
  for (const s of [-1, 1]) {
    const off = s * (side ? 1.2 : 3.1);
    const fwd = side ? legSwing * s * side : legSwing * s;
    g.fillStyle = s === 1 ? L.pants : shade(L.pants, -0.18);
    rr(g, cx + off - 2.6 + (side ? fwd * 0.6 : 0), top + 17, 5.2, 12 + (side ? 0 : Math.abs(fwd) * 0.1), 2.2);
    g.fill();
    // boot
    g.fillStyle = L.boots;
    rr(g, cx + off - 3 + (side ? fwd * 0.8 : 0), feet - 4, 6, 4.4, 1.8); g.fill();
    g.fillStyle = rgba('#fff', 0.12);
    g.fillRect(cx + off - 3 + (side ? fwd * 0.8 : 0), feet - 4, 6, 1);
  }

  // ---- torso
  const tw = side ? 9.5 : 12, th = 17;
  const tx = cx - tw / 2, ty = top + 2;
  const gr = g.createLinearGradient(tx, ty, tx + tw, ty + th);
  gr.addColorStop(0, shade(L.shirt, 0.2)); gr.addColorStop(0.55, L.shirt); gr.addColorStop(1, shade(L.shirt, -0.28));
  g.fillStyle = gr; rr(g, tx, ty, tw, th, 3); g.fill();
  g.strokeStyle = rgba('#000', 0.35); g.lineWidth = 0.8; g.stroke();

  if (L.coat) {   // lab coat overlay
    g.fillStyle = rgba('#f2f4f6', 0.95);
    rr(g, tx - 1, ty + 1, tw + 2, th + 4, 3); g.fill();
    g.strokeStyle = rgba('#b9bfc6', 0.9); g.lineWidth = 0.8; g.stroke();
    if (dir === 2) { g.strokeStyle = rgba('#aeb5bc', 0.9); g.beginPath(); g.moveTo(cx, ty + 2); g.lineTo(cx, ty + th + 3); g.stroke(); }
  }
  if (L.apron) {
    g.fillStyle = rgba('#6b4a2a', 0.9); rr(g, tx + 1, ty + 5, tw - 2, th - 3, 2); g.fill();
  }
  if (L.vest) {   // hi-vis vest with reflective bands
    g.fillStyle = rgba('#e8e24a', 0.92);
    rr(g, tx - 0.8, ty + 3, tw + 1.6, th - 3, 2.4); g.fill();
    g.strokeStyle = rgba('#8a8420', 0.8); g.lineWidth = 0.7; g.stroke();
    g.fillStyle = rgba('#d8dde2', 0.95);
    g.fillRect(tx - 0.8, ty + 7, tw + 1.6, 2);
    g.fillRect(tx - 0.8, ty + 12, tw + 1.6, 2);
    if (dir === 2) { g.fillStyle = rgba('#8a8420', 0.5); g.fillRect(cx - 0.5, ty + 3, 1, th - 3); }
  }
  if (L.tie && dir === 2) {
    g.fillStyle = '#a63a2e';
    g.beginPath(); g.moveTo(cx, ty + 1); g.lineTo(cx + 1.8, ty + 4); g.lineTo(cx, ty + 12); g.lineTo(cx - 1.8, ty + 4); g.closePath(); g.fill();
  }

  // ---- arms
  const armSwing = walk ? -swing * 4 : 0;
  const armY = ty + 3;
  for (const s of [-1, 1]) {
    let ax = cx + s * (tw / 2 + 1.2), ay = armY, len = 11, ang = 0;
    if (workPose >= 0) { ang = s * (workPose ? -0.9 : -0.5); ay -= 1; }
    else if (side) { ang = s === side ? armSwing * 0.12 : -armSwing * 0.12; }
    else ang = s * 0.08 + armSwing * 0.05 * s;
    g.save(); g.translate(ax, ay); g.rotate(ang);
    g.fillStyle = L.coat ? '#f2f4f6' : shade(L.shirt, s === 1 ? -0.1 : 0.08);
    rr(g, -2.1, 0, 4.2, len * 0.6, 2); g.fill();
    g.fillStyle = L.skin;             // forearm / hand
    rr(g, -1.9, len * 0.55, 3.8, len * 0.45, 1.8); g.fill();
    g.restore();
  }

  // ---- head
  const hy = top - 2.5, hr = 5.6;
  g.fillStyle = L.skin;
  g.beginPath(); g.ellipse(cx, hy, hr, hr * 1.06, 0, 0, TAU); g.fill();
  g.fillStyle = rgba('#000', 0.10);
  g.beginPath(); g.ellipse(cx + (side ? side * 1.6 : 0), hy + 1.6, hr * 0.9, hr * 0.7, 0, 0, TAU); g.fill();
  // neck shadow
  g.fillStyle = rgba('#000', 0.22); g.fillRect(cx - 2.2, hy + hr * 0.8, 4.4, 2);

  // hair
  if (!L.helmet || dir === 0) {
    g.fillStyle = L.hair;
    g.beginPath(); g.arc(cx, hy - 0.8, hr * 1.02, Math.PI, TAU); g.fill();
    if (L.long) { g.fillRect(cx - hr, hy - 1, hr * 2, 5); }
  }
  // face (only when facing camera or sideways)
  if (dir === 2) {
    g.fillStyle = '#2a2118';
    g.beginPath(); g.arc(cx - 2, hy + 0.4, 0.75, 0, TAU); g.fill();
    g.beginPath(); g.arc(cx + 2, hy + 0.4, 0.75, 0, TAU); g.fill();
    g.strokeStyle = rgba('#6b4636', 0.8); g.lineWidth = 0.7;
    g.beginPath(); g.arc(cx, hy + 2.6, 1.5, 0.2, Math.PI - 0.2); g.stroke();
    if (L.beard) { g.fillStyle = rgba(L.hair, 0.85); g.beginPath(); g.arc(cx, hy + 2.4, 3.4, 0.15, Math.PI - 0.15); g.fill(); }
    if (L.glasses) {
      g.strokeStyle = rgba('#1c1c20', 0.9); g.lineWidth = 0.8;
      g.beginPath(); g.arc(cx - 2, hy + 0.4, 1.9, 0, TAU); g.arc(cx + 2, hy + 0.4, 1.9, 0, TAU); g.stroke();
      g.fillStyle = rgba(P.glass, 0.3);
      g.beginPath(); g.arc(cx - 2, hy + 0.4, 1.8, 0, TAU); g.fill();
      g.beginPath(); g.arc(cx + 2, hy + 0.4, 1.8, 0, TAU); g.fill();
    }
  } else if (side) {
    g.fillStyle = '#2a2118';
    g.beginPath(); g.arc(cx + side * 2.4, hy + 0.4, 0.7, 0, TAU); g.fill();
  }

  // helmet
  if (L.helmet && dir !== 0) {
    const hg = g.createLinearGradient(cx - hr, hy - hr, cx + hr, hy);
    hg.addColorStop(0, shade(L.helmet, 0.32)); hg.addColorStop(0.6, L.helmet); hg.addColorStop(1, shade(L.helmet, -0.3));
    g.fillStyle = hg;
    g.beginPath(); g.arc(cx, hy - 0.6, hr * 1.08, Math.PI, TAU); g.fill();
    g.fillRect(cx - hr * 1.08, hy - 1.4, hr * 2.16, 1.8);
    // brim
    g.fillStyle = shade(L.helmet, -0.18);
    const bx = side ? cx + side * 1.5 : cx;
    rr(g, bx - hr * 1.1, hy - 0.4, hr * 2.2, 1.8, 0.9); g.fill();
    g.strokeStyle = rgba('#000', 0.4); g.lineWidth = 0.6; g.stroke();
    // crest rib
    g.fillStyle = rgba('#fff', 0.22);
    if (dir === 2) g.fillRect(cx - 0.7, hy - hr * 1.05, 1.4, hr * 0.9);
  } else if (L.helmet && dir === 0) {
    g.fillStyle = shade(L.helmet, -0.05);
    g.beginPath(); g.arc(cx, hy, hr * 1.08, 0, TAU); g.fill();
    g.fillStyle = rgba('#fff', 0.2);
    g.beginPath(); g.arc(cx - 1.6, hy - 1.6, hr * 0.4, 0, TAU); g.fill();
  }

  // ---- carried tool / prop
  if (L.tool && (workPose >= 0 || col === 0)) {
    const hx = cx + (side ? side * 7 : 7), hy2 = top + 13;
    if (L.tool === 'wrench') {
      g.save(); g.translate(hx, hy2); g.rotate(workPose === 1 ? -0.5 : 0.25);
      g.strokeStyle = P.steelHi; g.lineWidth = 2; g.lineCap = 'round';
      g.beginPath(); g.moveTo(0, 0); g.lineTo(0, -9); g.stroke();
      g.fillStyle = P.steelMid; g.beginPath(); g.arc(0, -10, 2.6, 0, TAU); g.fill();
      g.fillStyle = '#1a1c20'; g.beginPath(); g.arc(0, -10.6, 1.2, 0, TAU); g.fill();
      g.lineCap = 'butt'; g.restore();
    } else if (L.tool === 'box') {
      g.fillStyle = '#a5804a'; rr(g, cx - 6, top + 8, 12, 9, 1.5); g.fill();
      g.strokeStyle = '#4a3717'; g.lineWidth = 0.9; g.stroke();
      g.beginPath(); g.moveTo(cx - 6, top + 12.5); g.lineTo(cx + 6, top + 12.5); g.stroke();
    } else if (L.tool === 'tablet') {
      g.save(); g.translate(cx + (side ? side * 5 : 5), top + 12); g.rotate(-0.2);
      g.fillStyle = '#2a2f36'; rr(g, -3.5, -2.5, 7, 5, 1); g.fill();
      g.fillStyle = rgba(P.glowCold, 0.8); g.fillRect(-2.8, -1.9, 5.6, 3.8);
      g.restore();
    }
  }
}

/** Small 3/4 portrait used by dialogue and the employee roster. */
export function portrait(role, seed, size = 56) {
  const { c, g } = makeCanvas(size, size);
  const sheet = characterSheet(role, seed);
  g.imageSmoothingEnabled = true;
  const s = size / 18;
  g.save();
  g.beginPath(); rr(g, 0, 0, size, size, 6); g.clip();
  const bg = g.createLinearGradient(0, 0, 0, size);
  bg.addColorStop(0, '#333a44'); bg.addColorStop(1, '#1c2128');
  g.fillStyle = bg; g.fillRect(0, 0, size, size);
  g.drawImage(sheet.canvas, 2 * FW, 2 * FH, FW, FH, size / 2 - (FW * s) / 2, size * 0.08 - 2 * s, FW * s, FH * s);
  g.restore();
  g.strokeStyle = rgba('#0d0f12', 0.9); g.lineWidth = 2; rr(g, 1, 1, size - 2, size - 2, 6); g.stroke();
  g.strokeStyle = rgba(P.uiTrim, 0.6); g.lineWidth = 1; rr(g, 2.5, 2.5, size - 5, size - 5, 5); g.stroke();
  return c;
}
